In [3]:
import numpy as np
import pandas as pd

In [4]:
dataset = pd.read_excel(r"D:\DA\DA 02 tamrin\1_18150893050.xlsx")
dataset.to_csv("1_18150893050.csv")

read the file and covert it to csv

In [5]:
dataset = pd.read_csv(r"D:\DA\DA 02 tamrin\1_18150893050.csv")
data_array = dataset.to_numpy()


read the csv file and creat a numpy array

In [6]:
data_df = pd.DataFrame(data_array , columns=["id","InvoiceNo","StockCode","Description","Quantity","InvoiceDate","UnitPrice","CustomerID","Country"])

creat a data frame

In [7]:
print(data_df)

            id InvoiceNo StockCode                          Description  \
0            0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER   
1            1    536365     71053                  WHITE METAL LANTERN   
2            2    536365    84406B       CREAM CUPID HEARTS COAT HANGER   
3            3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE   
4            4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.   
...        ...       ...       ...                                  ...   
541904  541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS   
541905  541905    581587     22899         CHILDREN'S APRON DOLLY GIRL    
541906  541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL    
541907  541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE   
541908  541908    581587     22138        BAKING SET 9 PIECE RETROSPOT    

       Quantity          InvoiceDate UnitPrice CustomerID         Country  
0             6  2010-1

In [8]:
print(data_df.info())
print(data_df.isna().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           541909 non-null  object
 1   InvoiceNo    541909 non-null  object
 2   StockCode    541909 non-null  object
 3   Description  540455 non-null  object
 4   Quantity     541909 non-null  object
 5   InvoiceDate  541909 non-null  object
 6   UnitPrice    541909 non-null  object
 7   CustomerID   406829 non-null  object
 8   Country      541909 non-null  object
dtypes: object(9)
memory usage: 37.2+ MB
None
id                  0
InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64


in this file, there is 541909 samples.all are object. in two columns: Description,CustomerID there is 1454 and 135080 missing data

In [20]:
data_df["Quantity"] = pd.to_numeric(data_df["Quantity"], errors="coerce")
data_df["CustomerID"] = pd.to_numeric(data_df["CustomerID"], errors="coerce")
data_df["UnitPrice"] = pd.to_numeric(data_df["UnitPrice"], errors="coerce")
data_df['InvoiceDate'] = pd.to_datetime(data_df['InvoiceDate'])

change Quantity,CustomerID,UnitPrice type to float.

In [10]:
data_df["CustomerID"] = data_df["CustomerID"].fillna("unknown")

fill null datas in CustomerID to unknown.

In [11]:
data_df = data_df.dropna(subset=["Description"])

deleting samples which are null in Description column.

In [21]:
print(data_df.info())
print(data_df.isna().sum())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 540455 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   id           540455 non-null  object        
 1   InvoiceNo    540455 non-null  object        
 2   StockCode    540455 non-null  object        
 3   Description  540455 non-null  object        
 4   Quantity     540455 non-null  int64         
 5   InvoiceDate  540455 non-null  datetime64[ns]
 6   UnitPrice    540455 non-null  float64       
 7   CustomerID   406829 non-null  float64       
 8   Country      540455 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(5)
memory usage: 41.2+ MB
None
id                  0
InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133626
Country             0
dtype: int64


In [40]:
data_df[["id", "Quantity"]].where(data_df["Quantity"]>500).dropna()

id  Quantity
730        730     600.0
3917      3917     600.0
4850      4850    1824.0
4945      4945    2880.0
4946      4946    1400.0
...        ...       ...
537883  537883     960.0
540061  540061     698.0
540070  540070    1500.0
540071  540071    1200.0
540421  540421   80995.0

[423 rows x 2 columns]

all transactions with negative quantities

In [39]:
data_df[["id", "UnitPrice"]].where(data_df["UnitPrice"] == 0).dropna()

id  UnitPrice
6391      6391        0.0
6392      6392        0.0
7313      7313        0.0
9302      9302        0.0
13217    13217        0.0
...        ...        ...
535336  535336        0.0
536908  536908        0.0
538504  538504        0.0
538505  538505        0.0
538919  538919        0.0

[1061 rows x 2 columns]

transactions where the unit price is zero.

In [38]:
data_df[["id", "Description"]].where(data_df["Description"].str.contains("gift", case=False, na=False)).dropna()

id                        Description
30          30                   STARS GIFT TAPE 
712        712          PACK OF 6 BIRDY GIFT TAGS
836        836       PACK OF 6 SWEETIE GIFT BOXES
839        839          PACK OF 6 BIRDY GIFT TAGS
950        950  SMALL STRIPES CHOCOLATE GIFT BAG 
...        ...                                ...
541502  541502      50'S CHRISTMAS GIFT BAG LARGE
541676  541676     6 GIFT TAGS VINTAGE CHRISTMAS 
541794  541794          PACK OF 6 BIRDY GIFT TAGS
541813  541813     6 GIFT TAGS VINTAGE CHRISTMAS 
541859  541859        TEA TIME TEAPOT IN GIFT BOX

[5293 rows x 2 columns]

all products whose description includes the word "gift"

In [37]:
data_df[["id", "Country"]].where(data_df["Country"] == "Germany").dropna()

id  Country
1109      1109  Germany
1110      1110  Germany
1111      1111  Germany
1112      1112  Germany
1113      1113  Germany
...        ...      ...
541801  541801  Germany
541802  541802  Germany
541803  541803  Germany
541804  541804  Germany
541805  541805  Germany

[9495 rows x 2 columns]

all orders placed from Germany

In [ ]:
data_df.loc[1000:1999, ["id", "InvoiceDate"]].where(data_df["InvoiceDate"].dt.month == 12).dropna()

id         InvoiceDate
1000  1000 2010-12-01 12:43:00
1001  1001 2010-12-01 12:43:00
1002  1002 2010-12-01 12:43:00
1003  1003 2010-12-01 12:43:00
1004  1004 2010-12-01 12:43:00
...    ...                 ...
1995  1995 2010-12-01 14:34:00
1996  1996 2010-12-01 14:34:00
1997  1997 2010-12-01 14:34:00
1998  1998 2010-12-01 14:34:00
1999  1999 2010-12-01 14:34:00

[995 rows x 2 columns]

all orders in December (month 12) from rows 1000 to 1999.